extensions/bitwarden/src/utils.ts

```diff
@@ -43,3 +43,37 @@ export function getPasswordGeneratingArgs(options: PasswordGeneratorOptions): st
 }
 
 export const capitalise = (value: string) => value.charAt(0).toUpperCase() + value.slice(1);
+
+export function withKeywords(items: Item[]): Item[] {
```

Comment:
Instead of checking each time if the property is null, you should push it to the array, and filter null value at the end of the function.

Reply:
Thank you for the code review! I’ve gone ahead and added the undefined or null values into the array, before filtering them at the end, although I’m not sure how to do this for the for loop, or card number handling. Do you have any suggestions?

```typescript
export function withKeywords(items: Item[]): Item[] {
  return items.map((item: Item) => {
    const keywords = new Set<string | null | undefined>(item.name);
    if (item.card) {
      const { brand, number } = item.card;
      keywords.add(brand);
      if (number !== null) {
        // Similar to Bitwarden, use the last 5 digits if the card is Amex
        const isAmex = /^3[47]/.test(number);
        keywords.add(number.substring(number.length - (isAmex ? 5 : 4), number.length));
      }
    }
    keywords.add(item.login?.username);a
    if (item.login?.uris) {
      for (const uri of item.login.uris) {
        if (uri.uri !== null) {
          try {
            new URL(uri.uri).hostname.split(".").forEach(hostname => keywords.add(hostname));
          } catch (error) {
            // Invalid hostname
          }
        }
      }
    }
    // Unique keywords that are not null (probably quicker than doing a filter)
    keywords.delete(null)
    keywords.delete(undefined)
    item.keywords = [...keywords as Set<string>]
    return item;
  });
```

Reply:
I was thinking about a filter: `items.keywords = keywords.filter(keyword => !!keyword)`

Reply:
Instead of mutating the Bitwarden items, I think it would be better to define an `ExtractKeywords(item: Item) -> string[]` function.

You would then call it directly when building the <List.Item /> component

`<List.Item keywords={ExtractKeywords(item)}  .../>`

Reply:
Changed to use `.filter`:

```typescript
export function withKeywords(items: Item[]): Item[] {
  return items.map((item: Item) => {
    const keywords: (string | null | undefined)[] = [item.name];
    if (item.card) {
      const { brand, number } = item.card;
      keywords.push(brand);
      if (number !== null) {
        // Similar to Bitwarden, use the last 5 digits if the card is Amex
        const isAmex = /^3[47]/.test(number);  // <-- HERE
        keywords.push(number.substring(number.length - (isAmex ? 5 : 4), number.length));
      }
    }
    keywords.push(item.login?.username);
    if (item.login?.uris) {
      for (const uri of item.login.uris) {  // <-- HERE
        if (uri.uri !== null) {
          try {
            keywords.push(...new URL(uri.uri).hostname.split("."));
          } catch (error) {
            // Invalid hostname
          }
        }
      }
    }
    // Unique keywords
    let filteredKeywords: string[] = keywords.filter((keyword): keyword is string => !(keyword === null || keyword === undefined));
    item.keywords = [...new Set(filteredKeywords)];
    return item;
  });
}
```

I'm still not sure how to remove the if statements for the highlighted locations.

Reply:
Are you happy with this now, @pomdtr 😄